File: python/problems/leetcode/0021_merge_two_sorted_lists/solution.py
"""LeetCode 21: Merge Two Sorted Lists."""

from __future__ import annotations


class ListNode:
    __slots__ = ("val", "next")

    def __init__(self, val: int = 0, next_: ListNode | None = None) -> None:
        self.val = val
        self.next = next_


class Solution:
    def mergeTwoLists(self, list1: ListNode | None, list2: ListNode | None) -> ListNode | None:
        dummy = ListNode(0)
        cur = dummy
        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
            else:
                cur.next = list2
                list2 = list2.next
            cur = cur.next
        cur.next = list1 or list2
        return dummy.next


def to_list(n: ListNode | None) -> list[int]:
    out: list[int] = []
    while n:
        out.append(n.val)
        n = n.next
    return out


def from_list(a: list[int]) -> ListNode | None:
    d = ListNode(0)
    c = d
    for x in a:
        c.next = ListNode(x)
        c = c.next
    return d.next


if __name__ == "__main__":
    a = from_list([1, 2, 4])
    b = from_list([1, 3, 4])
    m = Solution().mergeTwoLists(a, b)
    assert to_list(m) == [1, 1, 2, 3, 4, 4]
    print("0021 OK")

File: cpp/problems/leetcode/0021_merge_two_sorted_lists/solution.cpp
// LeetCode 21
#include <bits/stdc++.h>
#include <cassert>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int v = 0, ListNode* n = nullptr) : val(v), next(n) {}
};

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode dummy(0);
        ListNode* cur = &dummy;
        while (list1 && list2) {
            if (list1->val <= list2->val) {
                cur->next = list1;
                list1 = list1->next;
            } else {
                cur->next = list2;
                list2 = list2->next;
            }
            cur = cur->next;
        }
        cur->next = list1 ? list1 : list2;
        return dummy.next;
    }
};

vector<int> to_vec(ListNode* n) {
    vector<int> o;
    for (; n; n = n->next) o.push_back(n->val);
    return o;
}

ListNode* from_vec(const vector<int>& a) {
    ListNode d(0);
    ListNode* c = &d;
    for (int x : a) {
        c->next = new ListNode(x);
        c = c->next;
    }
    return d.next;
}

int main() {
    ListNode* a = from_vec({1, 2, 4});
    ListNode* b = from_vec({1, 3, 4});
    ListNode* m = Solution().mergeTwoLists(a, b);
    assert(to_vec(m) == vector<int>({1, 1, 2, 3, 4, 4}));
    cout << "0021 OK" << endl;
    return 0;
}
